src/core/Velodrome.sol

```diff
@@ -26,87 +28,92 @@ interface IVelodromePair {
+        unchecked {
+            return _d(y, 3 * x0, x0 * x0 / _VELODROME_NEWTON_BASIS * x0 / _VELODROME_NEWTON_BASIS);
+        }
+    }
+
     function _d(uint256 y, uint256 three_x0, uint256 x0_cubed) private pure returns (uint256) {
         unchecked {
-            return _d(y, three_x0, x0_cubed, y * y / _BASIS);
+            return _d(y, three_x0, x0_cubed, y * y / _VELODROME_NEWTON_BASIS);
         }
     }
 
     function _d(uint256, uint256 three_x0, uint256 x0_cubed, uint256 y_squared) private pure returns (uint256) {
         unchecked {
-            return y_squared * three_x0 / _BASIS + x0_cubed;
+            return y_squared * three_x0 / _VELODROME_NEWTON_BASIS + x0_cubed;
         }
     }
 
     error NotConverged();
 
     // Using Newton-Raphson iterations, compute the smallest `new_y` such that `_k(x0, new_y) >=
     // xy`. As a function of `y`, we find the root of `_k(x0, y) - xy`.
-    function _get_y(uint256 x0, uint256 xy, uint256 y) private pure returns (uint256) {
+    function _get_y(uint256 x0, uint256 xy, uint256 y) internal pure returns (uint256) {
         unchecked {
             uint256 three_x0 = 3 * x0;
-            uint256 x0_squared = x0 * x0 / _BASIS;
-            uint256 x0_cubed = x0_squared * x0 / _BASIS;
+            uint256 x0_squared = x0 * x0 / _VELODROME_NEWTON_BASIS;
+            uint256 x0_cubed = x0_squared * x0 / _VELODROME_NEWTON_BASIS;
             for (uint256 i; i < 255; i++) {
-                uint256 y_squared = y * y / _BASIS;
+                uint256 y_squared = y * y / _VELODROME_NEWTON_BASIS;
                 uint256 k = _k(x0, y, x0_squared, y_squared);
+                uint256 d = _d(y, three_x0, x0_cubed, y_squared);
                 if (k < xy) {
-                    // there are two cases where dy == 0
-                    // case 1: The y is converged and we find the correct answer
-                    // case 2: _d(x0, y) is too large compare to (xy - k) and the rounding error
-                    //         screwed us.
-                    //         In this case, we need to increase y by 1
-                    uint256 dy = ((xy - k) * _BASIS).unsafeDiv(_d(y, three_x0, x0_cubed, y_squared));
-                    if (dy == 0) {
-                        if (k == xy) {
-                            // We found the correct answer. Return y
-                            return y;
-                        }
-                        if (_k(x0, y + 1, x0_squared) > xy) {
-                            // If _k(x0, y + 1) > xy, then we are close to the correct answer.
-                            // There's no closer answer than y + 1
-                            return y + 1;
-                        }
-                        dy = 1;
-                    }
+                    uint256 dy = (xy - k).unsafeMulDiv(_VELODROME_NEWTON_BASIS, d);
                     y += dy;
+                    if (dy <= _VELODROME_NEWTON_EPS) {
+                        return y + dy;
+                    }
```

Comment:
I got a bit confused here: You say that once we are within this bound we're good, but you're adding `dy` to `y` again right after updating the `y` (Effectively being `y + 2 * dy`). Is this just in case or something else? I feel like I'm missing some context.

Reply:
This was to account for rounding error in the computation of `k` in the AMM. I've replaced this code with something a bit more explicit and well-commented.